spec/classes/producer_spec.rb

```diff
@@ -2,78 +2,51 @@
-  end
-
-  it { is_expected.to contain_class('kafka::producer::install').that_comes_before('Class[kafka::producer::config]') }
-  it { is_expected.to contain_class('kafka::producer::config').that_comes_before('Class[kafka::producer::service]') }
-  it { is_expected.to contain_class('kafka::producer::service').that_comes_before('Class[kafka::producer]') }
-  it { is_expected.to contain_class('kafka::producer') }
-
-  context 'on Debian' do
-    describe 'kafka::producer::install' do
-      context 'defaults' do
-        it { is_expected.to contain_class('kafka') }
+  on_supported_os.each do |os, os_facts|
+    context "on #{os}" do
+      let(:facts) do
+        os_facts
       end
-    end
 
-    describe 'kafka::producer::config' do
-      context 'defaults' do
-        it { is_expected.to contain_file('/opt/kafka/config/producer.properties') }
+      let :params do
+        {
+          service_config: {
+            'broker-list' => 'localhost:9092',
+            'topic'       => 'demo'
+          },
+          input: '/tmp/kafka-producer'
+        }
       end
-    end
 
-    describe 'kafka::producer::service' do
-      context 'defaults' do
-        it { is_expected.to contain_file('/etc/init.d/kafka-producer') }
+      it { is_expected.to contain_class('kafka::producer::install').that_comes_before('Class[kafka::producer::config]') }
+      it { is_expected.to contain_class('kafka::producer::config').that_comes_before('Class[kafka::producer::service]') }
+      it { is_expected.to contain_class('kafka::producer::service').that_comes_before('Class[kafka::producer]') }
+      it { is_expected.to contain_class('kafka::producer') }
 
-        it { is_expected.to contain_service('kafka-producer') }
+      describe 'kafka::producer::install' do
+        context 'defaults' do
+          it { is_expected.to contain_class('kafka') }
+        end
       end
-    end
-  end
 
-  context 'on CentOS' do
-    let :facts do
-      {
-        osfamily: 'RedHat',
-        os: { family: 'RedHat' },
-        operatingsystem: 'CentOS',
-        operatingsystemrelease: '7',
-        operatingsystemmajrelease: '7',
-        architecture: 'amd64',
-        path: '/usr/local/sbin',
-        service_provider: 'systemd'
-      }
-    end
+      describe 'kafka::producer::config' do
+        context 'defaults' do
+          it { is_expected.to contain_file('/opt/kafka/config/producer.properties') }
+        end
+      end
 
-    describe 'kafka::producer::service' do
-      context 'defaults' do
-        it { is_expected.to raise_error(Puppet::Error, %r{Console Producer is not supported on systemd, because the stdin of the process cannot be redirected}) }
+      describe 'kafka::producer::service' do
+        context 'defaults' do
+          if os_facts[:service_provider] == 'systemd'
+            it { is_expected.to raise_error(Puppet::Error, %r{Console Producer is not supported on systemd, because the stdin of the process cannot be redirected}) }
```

Comment:
It was there before, but you can also use
```suggestion
            it { is_expected.to compile.and_raise_error(%r{Console Producer is not supported on systemd, because the stdin of the process cannot be redirected}) }
```